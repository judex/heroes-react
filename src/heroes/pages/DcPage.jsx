import { HeroList } from "../components"

export const DcPage = () => {
  return (
    <>
      <h1 className="font-black text-2xl mb-6">Dc Page</h1>
      <HeroList publisher="DC Comics" />
    </>
  )
}
