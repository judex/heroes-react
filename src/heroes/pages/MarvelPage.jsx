import { HeroList } from "../components/HeroList"

export const MarvelPage = () => {
  return (
    <>
      <div className="font-black text-2xl mb-6">Marvel Page</div>
      <HeroList publisher="Marvel Comics" />
    </>
  )
}
