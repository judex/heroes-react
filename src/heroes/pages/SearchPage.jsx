import queryString from "query-string";
import { useLocation, useNavigate } from "react-router-dom";
import { Notification } from "../../components/Notification";
import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../components";
import { getHeroesByName } from "../helpers";
export const SearchPage = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const { q = "" } = queryString.parse(location.search); // Extrae el valor de 'q' de la URL
  const heroes = getHeroesByName(q); // Obtiene los héroes que coinciden con la búsqueda

  const showNotification = q.length === 0; // Muestra la notificación si no hay valor de búsqueda
  const showNoResults = q.length > 0 && heroes.length === 0; // Muestra la notificación si no se encuentran héroes

  const { searchText, onInputChange, onResetForm } = useForm({
    searchText: q,
  });

  const onSearchSubmit = (event) => {
    event.preventDefault();
    if (searchText.trim().length <= 1) return;
    navigate(`?q=${searchText.toLowerCase().trim()}`); // Navega a la ruta con el query string
    //onResetForm(); // Reinicia el formulario después de enviar
  };

  return (
    <>
      <div className="font-black text-2xl mb-6">Searching Page</div>
      <form
        aria-label="form"
        autoComplete="off"
        className="max-w-md mx-auto"
        onSubmit={onSearchSubmit}
      >
        <label
          htmlFor="default-search"
          className="mb-2 text-sm font-medium text-gray-900 sr-only dark:text-white"
        >
          Search
        </label>
        <div className="relative">
          <div className="absolute inset-y-0 start-0 flex items-center ps-3 pointer-events-none">
            <svg
              className="w-4 h-4 text-gray-500 dark:text-gray-400"
              aria-hidden="true"
              xmlns="http://www.w3.org/2000/svg"
              fill="none"
              viewBox="0 0 20 20"
            >
              <path
                stroke="currentColor"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="2"
                d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"
              />
            </svg>
          </div>
          <input
            type="search"
            id="default-search"
            name="searchText"
            value={searchText}
            onChange={onInputChange}
            className="block w-full p-4 ps-10 text-sm text-gray-900 border border-gray-300 rounded-lg bg-gray-50 focus:ring-blue-500 focus:border-blue-500 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
            placeholder="Search Heroes ..."
            required
          />
          <button
            type="submit"
            className="text-white absolute end-2.5 bottom-2.5 bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-sm px-4 py-2 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800"
          >
            Search
          </button>
        </div>
      </form>
      <div className="mt-6">
        <h2 className="font-black text-2xl mb-6">Results</h2>

        {/* Notificaciones */}
        {showNotification && (
          <Notification type="warning" message="No search value entered" />
        )}

        {showNoResults && (
          <Notification type="error" message={`No results found for "${q}"`} />
        )}

        {/* Resultados */}
        {!showNotification &&
          !showNoResults &&
          heroes.map((hero) => <HeroCard key={hero.id} {...hero} />)}
      </div>
    </>
  );
};
