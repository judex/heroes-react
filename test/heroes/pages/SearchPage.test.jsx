import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

jest.spyOn(console, "warn").mockImplementation((message) => {
  if (!message.includes("React Router Future Flag Warning")) {
    console.warn(message);
  }
});

const mockUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockUseNavigate,
}));

describe("Pruebas en <SearchPage />", () => {
  beforeEach(() => jest.clearAllMocks()); // 100% limpiar los mocks antes de cada prueba
  const urlTemp = "batman";
  test("debe de mostrar el componente sin heroes que es por defecto", () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    );
    // screen.debug();
    expect(container).toMatchSnapshot();
    expect(screen.getByText("Searching Page")).toBeTruthy();
  });
  test("debe de mostrar a batman y el input con el valor del query string", () => {
    const { container } = render(
      <MemoryRouter initialEntries={[`/search?q=${urlTemp}`]}>
        <SearchPage />
      </MemoryRouter>
    );
    // screen.debug();
    // expect(container).toMatchSnapshot();
    const input = screen.getByRole("searchbox");
    expect(input.value).toBe(urlTemp);
    const img = screen.getByRole("img", { name: "Batman" });
    screen.debug(img);
    expect(img).toBeTruthy();
    expect(img.src).toContain("/heroes/dc-batman.jpg");
    const nroPalabras = urlTemp.length === 0;
    expect(nroPalabras).toBeFalsy();
    expect(screen.queryByText("No heroes found")).toBeFalsy();
  });

  test("debe de mostrar un error si no se encuentran el hero (batman123)", () => {
    const heroeSearch = "batman123";
    render(
      <MemoryRouter initialEntries={[`/search?q=${heroeSearch}`]}>
        <SearchPage />
      </MemoryRouter>
    );
    // screen.debug();
    expect(
      screen.getByText(`No results found for "${heroeSearch}"`)
    ).toBeTruthy();
  });

  test("debe de llamar el navigate a la ruta ", () => {
    const heroeSearch = "superman";

    render(
      <MemoryRouter initialEntries={["/search"]}>
        <SearchPage />
      </MemoryRouter>
    );
    screen.debug();
    const input = screen.getByRole("searchbox");
    // screen.debug(input);
    fireEvent.change(input, {
      target: { name: "searchText", value: heroeSearch },
    });

    const form = screen.getByRole("form");
    fireEvent.submit(form);

    expect(mockUseNavigate).toHaveBeenCalledWith(`?q=${heroeSearch}`);
  });
});
