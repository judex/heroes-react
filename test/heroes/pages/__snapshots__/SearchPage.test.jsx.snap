// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> debe de mostrar el componente sin heroes que es por defecto 1`] = `
<div>
  <div
    class="font-black text-2xl mb-6"
  >
    Searching Page
  </div>
  <form
    aria-label="form"
    autocomplete="off"
    class="max-w-md mx-auto"
  >
    <label
      class="mb-2 text-sm font-medium text-gray-900 sr-only dark:text-white"
      for="default-search"
    >
      Search
    </label>
    <div
      class="relative"
    >
      <div
        class="absolute inset-y-0 start-0 flex items-center ps-3 pointer-events-none"
      >
        <svg
          aria-hidden="true"
          class="w-4 h-4 text-gray-500 dark:text-gray-400"
          fill="none"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
      <input
        class="block w-full p-4 ps-10 text-sm text-gray-900 border border-gray-300 rounded-lg bg-gray-50 focus:ring-blue-500 focus:border-blue-500 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
        id="default-search"
        name="searchText"
        placeholder="Search Heroes ..."
        required=""
        type="search"
        value=""
      />
      <button
        class="text-white absolute end-2.5 bottom-2.5 bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-sm px-4 py-2 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800"
        type="submit"
      >
        Search
      </button>
    </div>
  </form>
  <div
    class="mt-6"
  >
    <h2
      class="font-black text-2xl mb-6"
    >
      Results
    </h2>
    <div
      class="flex items-center w-full p-4 rounded-lg shadow-sm bg-yellow-100 text-yellow-700 dark:bg-yellow-800 dark:text-yellow-200"
      role="alert"
    >
      <div
        class="ms-3 text-sm font-normal"
      >
        No search value entered
      </div>
    </div>
  </div>
</div>
`;
